### PR TITLE
Outfitter Shipyard Descriptions

### PR DESCRIPTION
**Pull request by rovermicrover**
## Feature Details
For https://github.com/endless-sky/endless-sky/issues/6652

Allow for Shipyard's and Outfitters to have descriptions. Either directly on planets via

```
planet "Foo"
    shipyard description "THIS IS A SHIPYARD YO!"
```

Or more generally on the Shipyard our Outfitter type

```
shipyard "ShipShip"
    description "This is a SHIPYARD maybe?!"
```

## UI Screenshots
image

## Testing Done
Tested on this [build](https://github.com/rovermicrover/endless-sky/releases/tag/out-sy-desc)

### Automated Tests Added
N/A

## Performance Impact
Just printing some more text.

### CODE REVIEW

**Comment by warp-core** [comment]
+1 to Zitchas' questions (I'm not reading all the code right now).

You should add support to remove these descriptions from both planets and sales, instead of exclusively appending to them.

You should move the example descriptions you're using to demonstrate the functionality into either an attached plugin or the `changes` section of an attached save file.

Above the info display for the currently selected item doesn't seem to me like a good place for this. It feels like shoving it in wherever for the sake of it. I would instead suggest using the space below where the selected ship sprite is displayed and only displaying the description when no item is selected.

File: data/map.txt [hunk 1]
```diff
@@ -30168,10 +30168,12 @@ planet Luna
 	shipyard "Navy Basics"
 	shipyard "Betelgeuse Basics"
 	shipyard "Syndicate Basics"
+	shipyard description "Don't call the Luna shipyard workers Lunatics. They seem to be very sensitive about it."
 	outfitter "Common Outfits"
 	outfitter "Lovelace Basics"
 	outfitter "Syndicate Basics"
 	outfitter "Ammo North"
+	outfitter description "Don't call the Luna outfitter workers Lunatics. They seem to be very sensitive about it."
```
Comment (Hurleveur) [inline data/map.txt:30176]: normally I'd say to not include data changes but this is too good to remove. keep it!

File: source/OutfitterPanel.cpp [hunk 2]
```diff
@@ -261,18 +261,33 @@ int OutfitterPanel::DetailWidth() const
...
 	const Color &bright = *GameData::Colors().Get("bright");
 	const Color &dim = *GameData::Colors().Get("medium");
 	const Sprite *collapsedArrow = SpriteSet::Get("ui/collapsed");
 
-	int heightOffset = 20;
+	int heightOffset = 0;
+
+	if(planet && !planet->OutfitterDescription().empty())
+	{
+		Point descriptionPoint(center.X() - INFOBAR_WIDTH / 2 + 25, center.Y() + heightOffset);
```
Comment (Hurleveur) [inline source/OutfitterPanel.cpp:274]: heightoffset is always 0 here?

File: source/Planet.cpp [hunk 3]
```diff
@@ -388,6 +392,22 @@ const Sale<Ship> &Planet::Shipyard() const
 
 
 
+// Get the shipyard description
+string Planet::ShipyardDescription() const
+{
+	if(!shipyardDescription.empty())
+		return shipyardDescription;
+
+	string shipyardDescriptionFromSales = "";
```
Comment (Hurleveur) [inline source/Planet.cpp:401]: I don't think you need to explicitely say its an empty string, just declaring it will make it be default I think

File: source/Planet.cpp [hunk 4]
```diff
@@ -408,6 +428,22 @@ const Sale<Outfit> &Planet::Outfitter() const
 
 
 
+// Get the outfitter description
+string Planet::OutfitterDescription() const
+{
+	if(!outfitterDescription.empty())
+		return outfitterDescription;
+
+	string outfitterDescriptionFromSales = "";
```
Comment (Hurleveur) [inline source/Planet.cpp:437]: same here
also maybe the outfitter and shipyard description should add up to each others.
Furthermore, this kind of info could be shown from the MapDetailPanel when the relevant thing is selected on the planet card? (like when the outfitter/shipyard is selected)

File: source/ShipyardPanel.cpp [hunk 5]
```diff
@@ -156,18 +156,33 @@ int ShipyardPanel::DetailWidth() const
...
+		description.SetFont(font);
+		description.Wrap(planet->ShipyardDescription());
+		heightOffset += description.Height();
+		description.Draw(descriptionPoint, bright);
+	}
+	// Draw this string representing the selected ship (if any), centered in the details side panel
+	Point selectedPoint(center.X() - INFOBAR_WIDTH / 2, center.Y() + heightOffset);
+	font.Draw({selectedItem, {INFOBAR_WIDTH - 20, Alignment::CENTER, Truncate::MIDDLE}},
+		selectedPoint, bright);
+	heightOffset += 20;
```
Comment (Hurleveur) [inline source/ShipyardPanel.cpp:181]: this all seems to be the same code as for the outfitterpanel... dont they both inherit of smth? mb not and if not you dont need to change it it's fine I guess

**Review by Hurleveur — CHANGES_REQUESTED** [review]
looks good, btw answer to warp's question is that the planet description has priority over other descriptions
otherwise shipyard/outfitter add up to each others

**Comment by Zitchas** [comment]
Alright, so if there's multiple descriptions, then it will show Planet description first, then each of the shipyard descriptions (presumably in the order the shipyards are listed in the planet specification?)?

Could there be an option under the planet description to supress shipyard descriptions? With some careful writing we can make them work nicely, but I would imagine that sometimes there may be the need to just outright supress shipyard descriptions.

Another suggestion: would it be possible to either (or both):
a) have the shipyard description disapear when the player clicks on a ship. (so it is only visible when the player initially visits the shipyard, not when they have a ship selected) and/or
b) Have the shipyard description have a small arrow next to it that enables it to be collapsed.

File: source/OutfitterPanel.cpp (hunk 2 above)
Comment (rovermicrover) [reply source/OutfitterPanel.cpp:274]: I added the offset to each initial point and then increment the offset after draw. That way the draw calls can be moved around without issue.

**Comment by rovermicrover** [comment]
Can easily make the description go away when a ship is selected. That seems like good behavior. Can also move it below the “selected title section”.

Will add the requested load behavior for overwriting, removing, etc.

And just to clarify the planet based descriptions for outfit and shipyard if present will be the only thing that shows up.

Otherwise the shipyard sale or outfitter sale descriptions will be appended together and displayed in a stable order based on the set ordering.

**Comment by rovermicrover** [comment]
@Zitchas 

As for the over all layout. What you are asking for is akin to web based responsive design. IE the columns would be in ratio to the over all display width. The text would also need to scale in some ratio.

That by itself sounds like a great issue, bordering on epic. A good resource for outline the issue would be the [Moz explanation of CSS flex box](https://developer.mozilla.org/en-US/docs/Learn/CSS/CSS_layout/Flexbox).

File: data/map.txt (hunk 1 above)
Comment (warp-core) [reply data/map.txt:30176]: I think that's exactly why it's not really a great addition here.

Not only have you used basically the same text in the shipyard and outfitter, but is the implication that someone has put signs up around the shipyard and outfitter that say this? It makes sense as a hail or even as a news item, but it seems a bit silly here.

And, the fact that it is copied from the hail and reused in both the shipyard and outfitter looks extremely lazy and low quality.

Comment (rovermicrover) [reply data/map.txt:30176]: Ok will move it to just the build branch I have and link that build tag here.

**Comment by Zitchas** [comment]
Sounds good, although if that's the case, then having an option for the planetary shipyard description to explicitly call the shipyard sections to be appended to it?

My thought is that in general, we'd have something like this: (assuming Geminus and Covert both had the same Lionsheart and Syndicate shipyards, which they don't)

Geminus Shipyard `This is a well built facility with plenty of sunshine where you can comfortably stroll amongst the ships. Several manufacturer booths are near the entrance.`
Shipyards:
- Lionsheart "display description"
- Syndicate "display description"
- Deep "no description"

Covert Shipyard `This dark and shadowy facility swarms with dark figures quietly stripping the markings off the ships coming in. Somehow, you doubt most of these ships were shipped here direct from the manufacturer.`
Shipyards:
- Lionsheart "no description"
- Syndicate "no description"
- Deep "no description"

Lionheart Shipyard `Lionheart's sales rep has a small display with samples to showcase their lightweight materials against typical materials, and the rep stands by eager to answer questions.`

Syndicate Shipyard `The Syndicate sales rep looks bored, sitting in an office adjacent to a large sign that declares "Your bottom line matters to us - when every credit counts, look to Syndicate Shipyards."`

So then when you go to Geminus, you get:

`This is a well built facility with plenty of sunshine where you can comfortably stroll amongst the ships. Several manufacturer booths are near the entrance.`
'Lionheart's sales rep has a small display with samples to showcase their lightweight materials against typical materials, and the rep stands by eager to answer questions.`
`The Syndicate sales rep looks bored, sitting in an office adjacent to a large sign that declares "Your bottom line matters to us - when every credit counts, look to Syndicate Shipyards."`

But when you go to Covert, you just get:
`This dark and shadowy facility swarms with dark figures quietly stripping the markings off the ships coming in. Somehow, you doubt most of these ships were shipped here direct from the manufacturer.`

So that way legit shipyards can use the shipyard lists to convey some information about the lists they contain, while others can skip it.

**Comment by Amazinite** [comment]
This doesn't seem like a good spot for a shop description to me. The right-most panel is the information for your currently selected ship, while the middle panel is the information for your currently selected item in the shop. Adding outfitter or shipyard descriptions above that seems odd to me. Even if the description only showed when nothing was selected, that still seems like a very odd place to put the shop description to me.

I'd say I'm not really sold on the idea in general, though. Why give descriptions to outfitters and shipyards? The linked issue says to make the outfitter and shipyard more like the spaceport, but shops and the spaceport have very different UI designed with very different needs in mind. That is to say, the spaceport was designed with the description location in mind. The shop has not. Therefore, without what would probably be some significant UI work to add a description without it looking out of place, I don't think we should add shipyard or outfitter descriptions.

**Comment by Zitchas** [comment]
Well, my initial concept/idea was that this description would appear in that side bar only when there wasn't a ship selected. So when first entering it, one gets the description instead of merely an empty column that says "no ship selected". This is probably still the preferable method.

**Comment by rovermicrover** [comment]
The description is now under the "select title" and goes away when something is selected.

Have also moved the title render, and the description render to method on the base class "ShopPanel"

**Review by Hurleveur — APPROVED** [review]
on the code side it looks for me, idk how it'd look on the content side though, I think that now it will only display if no ship is selected which sounds good to me, so long as it doesnt look weird

**Comment by rovermicrover** [comment]
Description now only appear when nothing is selected. And the add delete logic from else where is now supported.

File: source/Sale.h [hunk 6]
```diff
@@ -34,6 +34,8 @@ class Sale : public std::set<const Item *> {
 	void Add(const Sale<Item> &other);
 
 	bool Has(const Item *item) const;
+
+	std::string description;
```
Comment (RisingLeaf) [inline source/Sale.h:38]: Normally we have variables as private members that can be accesed through getters/setters.
I would prefer to have this here too.

**Review by RisingLeaf — CHANGES_REQUESTED** [review]
One small thing, otherwise looks good

File: source/Sale.h (hunk 6 above)
Comment (rovermicrover) [reply source/Sale.h:38]: Can do

**Review by RisingLeaf — APPROVED** [review]

**Comment by rovermicrover** [comment]
Working through a small issue where it now only looks at the outfitter description.

**Comment by rovermicrover** [comment]
Found the issue, verifying now

**Comment by rovermicrover** [comment]
Fix has been verified. I forgot to update one line after a refactor.

**Comment by Zitchas** [comment]
Looking forward to putting these into action. Thanks for working on this!

File: source/Sale.h [hunk 7]
```diff
@@ -55,11 +65,11 @@ void Sale<Item>::Load(const DataNode &node, const Set<Item> &items)
...
 			this->insert(items.Get(child.Token(1)));
 		else if(add && child.Size() >= 3 && child.Token(1) == "description")
-			description += child.Token(2) + '\n';
+			this->AppendDescription(child.Token(2) + '\n');
 		else if(remove && child.Size() >= 2 && child.Token(1) == "description")
-			description = "";
+			this->SetDescription("");
 		else if(token == "description" && child.Size() >= 2)
-			description += child.Token(1) + '\n';
+			this->AppendDescription(child.Token(1) + '\n');
```
Comment (Hurleveur) [inline source/Sale.h:72]: you dont need an appenddescription function, because description is a private member of sale already, and it wouldnt make sense for this to be possible from somewhere else

only the getdescription was useful and needed when making description private

Comment (rovermicrover) [reply source/Sale.h]: I read the above as a request for getters and setters for more encapsulation. I don't care either way they have been removed.

Comment (RisingLeaf) [reply source/Sale.h]: sorry, that is my fault

Comment (rovermicrover) [reply source/Sale.h]: NP. Have worked with people that take encapsulation very seriously before so thought nothing of it.

**Comment by UnorderedSigh** [comment]
You can use this save to test the feature. It has a mission that tells you to visit two planets to see the descriptions. Luna puts the descriptions in the planet definition, while Syndicate Advanced has it in the outfitter and shipyard definitions.

[Curt Shirt~3014-03-24 shipyard outfitter descriptions.txt](https://github.com/endless-sky/endless-sky/files/10777299/Curt.Shirt.3014-03-24.shipyard.outfitter.descriptions.txt)

**Review by UnorderedSigh — APPROVED** [review]
I can confirm that adding descriptions via events does work, both with this branch, and with the branch update to master (as of ten hours ago). I did not test removal or the numerous corner cases, though. However, adding is the most critical part.

**Comment by Amazinite** [comment]
As I described earlier, I don't think the the shops are good places to have descriptions. The purpose of shops is to view outfits and ships, and so the purpose of the UI should be entirely focused on that. I don't see the shops having their own descriptions as adding that purpose. If there is something unique about a planet's shipyard or outfitter, then that should probably be mentioned in the planet's description or spaceport description.
